sled-agent/src/sim/http_entrypoints_storage.rs

```diff
@@ -0,0 +1,169 @@
+        api.register(region_list)?;
+        api.register(region_create)?;
+        api.register(region_get)?;
+        api.register(region_delete)?;
+        Ok(())
+    }
+
+    let mut api = CrucibleAgentApiDescription::new();
+    if let Err(err) = register_endpoints(&mut api) {
+        panic!("failed to register entrypoints: {}", err);
+    }
+    api
+}
+
+// XXX XXX XXX THIS SUCKS XXX XXX XXX
+//
+// I need to re-define all structs used in the crucible agent
+// API to ensure they have the traits I need. The ones re-exported
+// through the client bindings, i.e., crucible_agent_client::types,
+// don't implement the "JsonSchema" trait, and cannot be used as
+// parameters to these Dropshot endpoints.
+//
+// I'd like them to! If we could ensure the generated client
+// also implemented e.g. JsonSchema, this might work?
+
+// To quickly test the type compatibility, uncomment the lines below,
+// and remove the hand-rolled implementations.
+
+// pub type RegionId = crucible_agent_client::types::RegionId;
+// pub type State = crucible_agent_client::types::State;
+// pub type CreateRegion = crucible_agent_client::types::CreateRegion;
+// pub type Region = crucible_agent_client::types::Region;
+// pub type RegionPath = crucible_agent_client::types::RegionPath;
+
+#[derive(
+    Serialize,
+    Deserialize,
+    JsonSchema,
+    Debug,
+    PartialEq,
+    Eq,
+    Clone,
+    PartialOrd,
+    Ord,
+)]
+pub struct RegionId(pub String);
+
+#[derive(Serialize, Deserialize, JsonSchema, Debug, PartialEq, Clone)]
+#[serde(rename_all = "lowercase")]
+pub enum State {
+    Requested,
+    Created,
+    Tombstoned,
+    Destroyed,
+    Failed,
+}
+
+#[derive(Serialize, Deserialize, JsonSchema, Debug, PartialEq, Clone)]
+pub struct CreateRegion {
+    pub id: RegionId,
+    pub volume_id: String,
+
+    pub block_size: u64,
+    pub extent_size: u64,
+    pub extent_count: u64,
+}
+
+#[derive(Serialize, Deserialize, JsonSchema, Debug, PartialEq, Clone)]
+pub struct Region {
+    pub id: RegionId,
+    pub volume_id: String,
+
+    pub block_size: u64,
+    pub extent_size: u64,
+    pub extent_count: u64,
+
+    pub port_number: u16,
+    pub state: State,
+}
+
+#[derive(Deserialize, JsonSchema)]
+struct RegionPath {
+    id: RegionId,
+}
+
+#[endpoint {
+    method = GET,
+    path = "/crucible/0/regions",
+}]
+async fn region_list(
+    rc: Arc<RequestContext<Arc<CrucibleData>>>,
+) -> Result<HttpResponseOk<Vec<Region>>, HttpError> {
+    let crucible = rc.context();
+    Ok(HttpResponseOk(crucible.list().await))
+}
+
+#[endpoint {
+    method = POST,
+    path = "/crucible/0/regions",
+}]
+async fn region_create(
+    rc: Arc<RequestContext<Arc<CrucibleData>>>,
+    body: TypedBody<CreateRegion>,
+) -> Result<HttpResponseOk<Region>, HttpError> {
+    let params = body.into_inner();
+    let crucible = rc.context();
+
+    Ok(HttpResponseOk(crucible.create(params).await))
+}
+
+#[endpoint {
+    method = GET,
+    path = "/crucible/0/regions/{id}",
```

Comment:
Is the `id` here, the UUID of the region?

Reply:
Yup! This format should match the server [within Crucible](https://github.com/oxidecomputer/crucible/blob/7e8d0c314adb43698e8a65d1c55abf2225d35c2c/agent/src/server.rs#L61-L65).